### PR TITLE
[SPO] Add unique list permissions support

### PR DESCRIPTION
**Pull request by timgrein**
## Closes https://github.com/elastic/enterprise-search-team/issues/5363

This PR adds support for unique site list permissions, which broke the inheritance to their parent site. This is guarded through a new RFC `fetch_unique_list_permissions`.

Also added some missing tests around `site_lists` and extended the `AsyncIterator` with some new assertions (which are needed for the `site_lists` tests) and added missing tests for this class, too.

Note: a separate PR will be opened for the Kibana RFC

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)

### CODE REVIEW

File: tests/commons.py [hunk 1]
```diff
@@ -1,3 +1,8 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
```
Comment (seanstory) [inline tests/commons.py:5]: Huh. We probably need a test to ensure that this appears on all our files. 
Nice catch.

File: tests/commons.py [hunk 2]
```diff
@@ -21,10 +27,18 @@ async def __anext__(self):
...
         if kwargs:
             self.call_kwargs.append(kwargs)
 
         return self
+
+    def assert_not_called(self):
+        return self.call_count == 0
+
+    def assert_called_once(self):
+        return self.call_count == 1
```
Comment (seanstory) [inline tests/commons.py:44]: I think this capability exists already: https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once

File: tests/commons.py [hunk 3]
```diff
@@ -21,10 +27,18 @@ async def __anext__(self):
...
         if args:
             self.call_args.append(args)
 
         if kwargs:
             self.call_kwargs.append(kwargs)
 
         return self
+
+    def assert_not_called(self):
+        return self.call_count == 0
```
Comment (seanstory) [inline tests/commons.py:41]: It baffles me that this isn't part of pytest. I've run across https://stackoverflow.com/questions/54838354/python-how-can-i-assert-a-mock-object-was-not-called-with-specific-arguments before. Lots of ways to hack around it I guess.

File: tests/test_commons.py
Comment (seanstory) [inline tests/test_commons.py:1]: Good ol' meta-tests. :D

**Review by seanstory — APPROVED** [review]

File: tests/commons.py (hunk 2 above)
Comment (timgrein) [reply tests/commons.py:44]: Yes but not on our custom `AsyncIterator` (which should probably be called `AsyncGenerator` class :) )

File: tests/commons.py (hunk 1 above)
Comment (timgrein) [reply tests/commons.py:5]: I think we check it somewhere for _some_ files. Maybe the `tests` directory is ignored?

**Pull request merged**